### ISSUE TITLE
Canonical Package Metadata

### ISSUE THREAD

**Issue by vyzo** [vyzo/gxpkg-daemon#2]
Package metadata can be implicit (deduced from the repo) or explicitly defined in gerbil.pkg (in the meta field).

**Comment by vyzo** [comment]
cc @belmarca @fare

**Comment by belmarca** [comment]
Ideally we minimise the amount of fields that have to be set by the authors.

Here is the current implementation's fields:

`author` is retrieved from the GitHub URL.
`name` is retrieved from the GitHub URL. No check is performed to check if it matches the `package:` field in `gerbil.pkg`.
`description` is retrieved from `gerbil.pkg`
`license` is retrieved from GitHub's API.
`runtime` is retrieved from `gerbil.pkg`.
`repo` is the submitted GitHub URL during registration.
`forks` is retrieved from GitHub's API.

Thus `description` and `runtime` are the only metadata fields definable by the author in `gerbil.pkg`. 

The fields `author`, `name`, `license`, `repo` and `forks` are computed.

**Comment by vyzo** [comment]
I am fully in agreement that we should maximally use implicit metadata from the hosting provider.

Some amenities we can get from github:
- description
- topics
- license
- origin (for forks)
- forks
- author
- stars

**Comment by belmarca** [comment]
Vendor specific metadata will be harder to normalise if we support GitHub, GitLab, etc.

This should probably be reserved for the package repository's web UI.

And I believe the `description`, at the very least a one-liner, should be a property inside `gerbil.pkg`. That facilitates sharing of libraries by email for example. Maybe we should also force the use of `author` inside `gerbil.pkg` for the same reason. An author might not use GitHub yet his package can be mirrored there by another person.

**Comment by vyzo** [comment]
What we can do is parse the `gerbil.pkg` metadata, and fill the missing values as fallback from the vendor.

I think we should have both `description` and `topics` as such optional metadata.
These two can aid search and categorization of packages a lot.

**Comment by vyzo** [comment]
And yeah, let's use `author` in the same way (with optional fallback).

**Comment by vyzo** [comment]
Also let's allow `authors` as an alternative to `author`.

**Comment by belmarca** [comment]
For topics, how do you deal with `topic1` vs `Topic1` vs `topic_1` vs `tpic1`? Do you have a pre-approved list of topics? Or anything goes? I don't want to manually clean up the topic list.

As for the `authors` key, sure, makes sense. We could also allow each author to define their full name, short name and email address, gpg key or whatever: e.g. `(author: '(github: belmarca full-name: "Marc-André Bélanger" email: marc-andre.belanger@asciimail.com)`.

If we define a set of mandatory keys, they need to be there or the package is invalid. I don't want to write `if`s to accommodate lazy programmers. That is why we need to keep the mandatory list at a minimum.

---

For the reasons I have described above, here is a revised list of _required_ metadata to be found in `gerbil.pkg` in order for a package to be considered a valid `gerbil` package by the metadata server. The metadata in `gerbil.pkg` is considered canonical and has precedence over information retrieved from providers such as GitHub.

`author(s)` is one or more strings.
`description` is a one line description of the package.
`license` is the short name of the license in use.

Here are examples of proposed valid `gerbil.pkg`:

```
(package: my-package
 author: myGitHubUsername
 description: "A nice package."
 license: "GPLv3")
```

```
(package: my-package
 author: "My full name"
 description: "A very nice package."
 license: "MIT")
```

```
(package: my-package
 authors: myGitHubUsername yourGitHubUsername "Another author's full name"
 description: "A very, very nice package."
 license: "GPLv3")
```

**Comment by vyzo** [comment]
I don't think we should reject packages because of missing metadata, especially if the server can fill them from the hosting provider.

@fare thoughts?

**Comment by vyzo** [comment]
Also, I think we should put the metadata in a `meta:` property in the plist, so that it can be extracted directly.
The package directive is meaningful for the package namespace, it's not quite metadata.

**Comment by belmarca** [comment]
Sure we can use a `meta:` property.

Also it would be relatively easy to add a `new` or `create` functionality to `gxpkg` that would scaffold a new package and ask the user for these required fields. That way it is standardized if we wish to force certain properties to be defined.

**Comment by fare** [comment]
For inspiration, in the Haskell world, the [Stackage](https://www.stackage.org/) guys have their shit together (whereas e.g. Hackage does the wrong thing with respect to dependency version constraints).

A few things:
  * `maintainer` should be a field separate from `author`
  * `description` too can be autodetected from github
  * `homepage` can be a useful url field.
  * `bug-tracker` is also an essential field, with a url for either a website or a mailing-list where to report bugs.
  * `mailto` for a discussion mailing-list, though maybe there should instead be two separate URL fields for `help-forum` and `developer-forum` or some such.
  * maybe support for repositories other than git, somehow?

**Comment by fare** [comment]
Also, will you validate the type of data in each of those fields?

**Comment by vyzo** [comment]
Well, we can validate the data for the fields, but not sure if we should reject packages or simply ignore malformed fields. We expect them to be strings, but some things are tricky to validate further (eg: mailto and urls).

**Comment by fare** [comment]
Definitely reject a field if of the wrong type. If there is a URL parser, even minimal, we should use it where appropriate.

**Comment by vyzo** [comment]
So there is also the idea floated by @belmarca, where we could have a `gxpkg init` command invocation to create and initialize gerbil.pkg with interactive metadata entry.

This could add validation at initialization time, as it's a little uncomfortable to have to reject valid packages in the server because of invalid  metadata.

**Comment by belmarca** [comment]
Note that this is why I am all for a required but minimal set of fields. It's hard to mess up and thus we theoretically shouldn't reject any packages (if you can write a library... you can fill in 3 metadata tags).
